cpp/src/tracing.cpp

```diff
@@ -0,0 +1,240 @@
+
+    // extract context from AMQP message annotations
+    const AMQPTextMapCarrier<proton::message::annotation_map> carrier(
+        &message.message_annotations());
+
+    nostd::shared_ptr<
+        opentelemetry::v0::context::propagation::TextMapPropagator>
+        prop = opentelemetry::context::propagation::GlobalTextMapPropagator::
+            GetGlobalPropagator();
+    context::Context current_ctx =
+        opentelemetry::context::RuntimeContext::GetCurrent();
+
+    context::Context new_context = prop->Extract(carrier, current_ctx);
+    options.parent =
+        opentelemetry::trace::propagation::GetSpan(new_context)->GetContext();
+
+    binary tag_in_binary = d.tag();
+    std::string tag_in_string = std::string(d.tag());
+    std::stringstream ss;
+    for(int i=0; i<(int)tag_in_string.length(); ++i)
+        ss << std::hex << (int)tag_in_binary[i];
+    std::string delivery_tag = ss.str();
+
+    receiver r = d.receiver();
+    source s = r.source();
+    std::string s_addr = s.address();
+
+    transfer tt(d);
+    std::string delivery_state = to_string(tt.state());
+
+    nostd::shared_ptr<trace::Span> span =
+        get_tracer()->StartSpan("amqp-message-received",
+                                {{"Delivery_tag", delivery_tag},
+                                 {"Source_address", s_addr}},
+                                options);
+
+    trace::Scope scope = get_tracer()->WithActiveSpan(span);
+
+    if (delivery_state == "unknown")
+    {
+        span->AddEvent("Delivery state is unknown");
+    }
+    if (delivery_state == "Received")
+    {
+        span->AddEvent("Delivery is received");
+    }
+
+    tag_span[tag_in_binary] = span;
+}
+
+void on_message_end_span(delivery &d) {
+    binary tag = d.tag();
+    nostd::shared_ptr<trace::Span> span = tag_span[tag];
+
+    span->End();
+}
+
+void on_settled_span(delivery &d) {
+    binary tag = d.tag();
+    nostd::shared_ptr<trace::Span> span = tag_span[tag];
+
+    span->End();
+}
+
+void send_span(message &message, binary &tag, tracker &track) {
+
+    opentelemetry::trace::StartSpanOptions options;
+    options.kind = opentelemetry::trace::SpanKind::kProducer;
+
+    binary tag_in_binary = tag;
+    std::string tag_in_string = std::string(tag);
+    std::stringstream ss;
+    for (int i = 0; i < (int)tag_in_string.length(); ++i)
+        ss << std::hex << (int)tag_in_binary[i];
+    std::string delivery_tag = ss.str();
+
+    sender s = track.sender();
+    target t = s.target();
+    std::string t_addr = t.address();
+
+    transfer tt(track);
+    std::string delivery_state = to_string(tt.state());
+
+    nostd::shared_ptr<trace::Span> span = get_tracer()->StartSpan(
+        "amqp-delivery-send",
+        {{"Delivery_tag", delivery_tag}, {"Destination_address", t_addr}},
+        options);
+
+    trace::Scope scope = proton::get_tracer()->WithActiveSpan(span);
+
+    if (delivery_state == "unknown") {
+        span->AddEvent("Delivery state is unknown");
+    }
```

Comment:
The send state will essentially always be unknown at this point. It becomes known when the disposition arrives from peer (e.g broker) setting the applicable state (for the messages that are not sent pre-settled).

Wherever the message 'tracker' is being updated (perhaps [1]) upon disposition is the point the delivery state will become known for send, and where the send span should be ended.

For pre-settled sends that may be different, as they will never get updated state due to being settled in advance. Depending on handling of the events, their spans may even need completed within the send call itself.

[1] https://github.com/apache/qpid-proton/blob/0.35.0/cpp/src/messaging_adapter.cpp#L157-L177

Reply:
@gemmellr  Ah, now it makes sense to me. I've made few changes, please let me know if I get it right.

Now I am ending the send_span where the message tracker is being updated and before ending; adding delivery state as an event (Now send_span delivery state is "accepted" for the demo file).